File: 오우영/Programmers/Level2/뒤에있는큰수찾기.js
/* 프로그래머스 - 뒤에 있는 큰 수 찾기 */

function solution(numbers) {
  let answer = new Array(numbers.length).fill(-1);
  let stack = [];
  for (let i = 0; i < numbers.length; i++) {
    while (stack && numbers[stack.at(-1)] < numbers[i]) {
      answer[stack.pop()] = numbers[i];
    }
    stack.push(i);
  }
  return answer;
}

File: 장세훈/Programmers/Level2/뒤에있는큰수찾기.java
class 뒤에있는큰수찾기 {
    /*
        https://school.programmers.co.kr/learn/courses/30/lessons/154539
     */
    public int[] solution(int[] numbers) {
        Deque<Integer> stack = new ArrayDeque<>();
        int[] answer = new int[numbers.length];
        Arrays.fill(answer, -1);

        for (int i = 0; i < numbers.length; i++) {
            while (!stack.isEmpty() && numbers[stack.peekLast()] < numbers[i]) {
                answer[stack.pollLast()] = numbers[i];
            }

            stack.addLast(i);
        }

        return answer;
    }
}